### DISCUSSION TITLE
Issues with 2 camera's on M8P with CM4 & CB1 - crowsnest#109

### DISCUSSION THREAD

**Discussion by mryel00** [mainsail-crew/mainsail#1833] [General]
This Discussion should be a place to discuss about https://github.com/mainsail-crew/crowsnest/issues/109.
Feel free to talk about this issue or ask for help about possible brought up solutions.

_The mainsail-crew will not provide any help, if this is a hardware issue (as we currently think) and we will only provide the platform for the discussion._
_We would be grateful if our community can help each other to find a good solution even for the non-tech-savvy users._

One found maybe possible [solution](https://github.com/orgs/mainsail-crew/discussions/1833#discussioncomment-8958956) and [explanation](https://github.com/orgs/mainsail-crew/discussions/1833#discussioncomment-8958750) of that issue.

**Comment by mryel00** [comment]
> We are going to run a bunch of comprehensive tests tomorrow morning to get empirical results when using the same cameras with different setups. Will let you know what we find.

@looxonline Sry to bother you, but did you guys already run those tests? I'm very interested in the results.

**Reply by looxonline** [reply]
Yup we did. Check it out:

TLDR:

- It is 100% a limit of the USB controller bandwidth combined with the cameras completely over reporting their needed bandwidth.
- For some reason the RPi cam driver reports a corrupt bandwidth requirement which we believe results in the kernel module deciding that it can still fit more cameras on the bus.
- When using the BTT Pi we could run 4 x 1080p cams in parallel because each USB port actually connects to its own internal controller. The total USB bandwidth on the BTT Pi is actually insane. A result that I never appreciated before running these tests.
- The CM4/CB1 both only expose one USB controller and then require it to be split through a hub so these both have the same bandwidth limitations of the RPi with the difference that the CB1 kernel camera drivers report a non-corrupt bandwidth requirement and therefore are limited to one cam.
- Counter intuitively, if mainsail team adds the option to run ustreamer in YUYV mode, we will be able to stream two cams in parallel even though YUYV is not compressed and MJPG is.

The long version:

Our test setup was:
1 x M8P
1 x CB1
1 x CM4
1 x BTT Pi
1 x RPi 4
1 x RPi 3B+

The CM4/CB1 were both connected to the M8P.

Test method:

- Connect a single camera to a USB port and ensure that we can get 30fps 1080p using MJPEG
- Send cat /sys/kernel/debug/usb/devices | grep "B: " and view the % bandwidth being requested on the controller for the single cam. If higher than 50% then the controller will reject a second cam.
- Attempt to stream a second cam and confirm the expected results.
- Change stream type to YUYV and enable the kernel module quirk that tells the kernel to ignore what the camera tells it and accurately calculate the bandwidth requirements. This only works with uncompressed streams.
- Repeat cat /sys/kernel/debug/usb/devices | grep "B: and check if the bandwidth on the bus has reduced.
- Connect two cams again and see if they will run.

Results:

- CB1 and CM4 have no problem on an M8P with a single, high res cam:

![image](https://github.com/mainsail-crew/mainsail/assets/19784349/9492cae5-f2c9-466b-85f4-c8ce18e82d7c)

- BTT Pi is the only device capable of running four, high res cameras at once since each USB port runs to its own controller with 480Mbps per controller.

![image](https://github.com/mainsail-crew/mainsail/assets/19784349/a50cc36d-863b-4d51-8485-222e396ca5a9)

- As expected, if we add a HUB to the BTT Pi on a single USB port and try to connect two cameras to the HUB, it fails. This is because a single camera reports > 50% required bus usage.
- Changing from MJPEG to YUYV and applying the kernel module quirk dramatically reduces the bus bandwidth use per camera and allows more than one on a high speed bus.
![image](https://github.com/mainsail-crew/mainsail/assets/19784349/e3188d57-1a63-4183-b654-b34c178a6b76)
- Another example of how the quirk allows us to get the kernel to see the actual required bus use.

![image](https://github.com/mainsail-crew/mainsail/assets/19784349/4178e033-3398-48bb-9844-b879956ad9f8)

- Ustreamer can be a PITA and often tries to max out the FPS in spite of what is configured. The image below shows how it jumps to something different to what is configured, almost maxing out what it thinks the bus can handle for the given res. This may cause frustration as users try to drop the FPS but find that it makes no impact to how many cameras they can run.

![image](https://github.com/mainsail-crew/mainsail/assets/19784349/20737335-7eff-48dd-a681-e049f22abb5e)

- RPi 3 camera drivers are drunk and report whack bus requirements, probably resulting in the kernel logic allowing more than one cam on the bus:

![image](https://github.com/mainsail-crew/mainsail/assets/19784349/6181c37e-1a63-47b6-b963-8a76119b4de6)

**Reply by mryel00** [reply]
Amazing work 😄 
These results are pretty interesting, especially that the Pi3 is reporting some weird stuff. Maybe some kernel modifications?

> Counter intuitively, if mainsail team adds the option to run ustreamer in YUYV mode, we will be able to stream two cams in parallel even though YUYV is not compressed and MJPG is.

Overall we already support that with `custom_flags`. Someone can add `--format=YUYV` and it will override the "forced" `MJPG` format.

> Changing from MJPEG to YUYV and applying the kernel module quirk dramatically reduces the bus bandwidth use per camera and allows more than one on a high speed bus.

So it's enabling `YUYV` and the quirk, so there isn't an easy way to fix it directly during runtime of crowsnest, if I understood that correctly. We shouldn't reload the uvc kernel module on start of crowsnest, as it might have some side effects for some specific setups.
We will discuss it internally if we can apply that quirk during install on the CB1 & CM4 and then set force `--format=YUYV` for those 2.

**Reply by looxonline** [reply]
Yes it must be both YUYV and the quirk. The quirk only applies to uncompressed streams unfortunately. In the other post I showed a kernel driver mod that would cause it to apply to compressed streams but that is far beyond the scope of most users.

I think it would be a good idea to force the quirk regardless of whether those modules are being used. Since the quirk only applies to uncompressed streams and almost all users use MJPEG it would have no effect on them and for those using YUYV, it will only have a positive effect since it does not under estimate the needed bandwidth, it just calculates it correctly.

By doing this, you can reduce the problem down to only needing to force YUYV on those modules and not have to worry about when or if to apply the quirk.

**Reply by mryel00** [reply]
> I think it would be a good idea to force the quirk regardless of whether those modules are being used.

We will discuss it internally, on what scope makes sense.
I hesitate a bit on applying it to every machine, even if it doesn't harm, as some users might not like it, if we apply it without their consent and a popup to ask might confuse others.

Anyway it won't come before the next big update, as we are currently rewriting everything of crowsnest and only apply a few last bug fixes atm to the current version.

Also it's good to know that it's actually more like a "bug" with the "normal" Raspberry Pis that they report wrong stuff, as this issue then might occur on other machines too. Actually another good reason to apply it overall, ig...

**Reply by mryel00** [reply]
Another small detail that I forgot in my previous response:
> Ustreamer can be a PITA and often tries to max out the FPS in spite of what is configured. The image below shows how it jumps to something different to what is configured, almost maxing out what it thinks the bus can handle for the given res. This may cause frustration as users try to drop the FPS but find that it makes no impact to how many cameras they can run.

Ustreamer sets the fps to what the camera is capable of. The camera can provide different encoded streams at specific resolutions with specific fps. Those can be shown with `v4l2-ctl -d <device> --list-ctrls-menus`. Those are fixed and nothing ustreamer can change. Those calues are different from cam to cam. A small snippet as an example of the output:
```
[0]: 'MJPG' (Motion-JPEG, compressed):
        Size: Discrete 2560x1440
                Interval: Discrete 0.033s (30.000 fps)
        Size: Discrete 1920x1080
                Interval: Discrete 0.033s (30.000 fps)
        Size: Discrete 640x480
                Interval: Discrete 0.017s (60.000 fps)
....
        Size: Discrete 1600x1200
                Interval: Discrete 0.033s (30.000 fps)
        Size: Discrete 960x720
                Interval: Discrete 0.033s (30.000 fps)
        Size: Discrete 1280x960
                Interval: Discrete 0.033s (30.000 fps)
[1]: 'YUYV' (YUYV 4:2:2):
        Size: Discrete 2560x1440
                Interval: Discrete 0.500s (2.000 fps)
        Size: Discrete 1920x1080
                Interval: Discrete 0.500s (2.000 fps)
        Size: Discrete 640x480
                Interval: Discrete 0.033s (30.000 fps)
....
        Size: Discrete 1600x1200
                Interval: Discrete 0.200s (5.000 fps)
        Size: Discrete 960x720
                Interval: Discrete 0.067s (15.000 fps)
        Size: Discrete 1280x960
                Interval: Discrete 0.200s (5.000 fps)
```
As you can see it differs from resolution to resolution what the cam can provide and it's different for the different encodings too. As `YUYV` isn't compressed you see as low as 2 fps. Those values shown, are the only combinations that will work, and you cannot enforce e.g. `1920x1080@20fps` with `MJPG` (maybe with some kernel modifications, but this is out of scope of this topic)

**Reply by looxonline** [reply]
The thing is that we had already confirmed that the FPS that we wanted were available at the resolutions that we had selected using the v4l2 compatibility list. It was really frustrating because it kept jumping to another even when the one we wanted was listed and should have worked.

**Comment by mryel00** [comment]
As a small summary to the proposed solution:
- Enable following quirk for the Linux UVC Driver: https://stackoverflow.com/a/25619508 (To survive reboots look at the answer below that)
- Use `custom_flags: --format=YUYV` for `mode: ustreamer` and `custom_flags: --camera-format=YUYV` for `mode: camera-streamer` inside your crowsnest.conf

After those two steps, you should be able to use 2 cams on your CB1 and CM4.
Thanks again for providing a solution, the long investigation and the good tests of @looxonline and his team.

If someone got questions or need help, feel free to ask in this discussion 😄 

P.S. The FPS with `YUYV` can be a lot lower than with `MJPG`. This differs from cam to cam.

**Reply by dragz** [reply]
Hi, and thanks for all the work you all put into this. I just tested the proposed solution and have made the following observations:

(0. I already have a working solution with two cams using camera-streamer on a M8P board with a CM4. Resolution has to set at 640x480 and fps to 10 on both cams.)
1. The suggested config does not improve on the situation for me. Switching to ustreamer with --format=YUYV and setting the uvcvideo param quirks=640 or 128 still does not let me increase the resolution or fps.
2. The suggested solution creates a lot more load on the cpu, 30-50%, just running the setup without printing. Using camera-streamer the cpu-load stays below 5%.

I´d be happy to do more testing if there are other suggestions to try out.

**Reply by mryel00** [reply]
If I understood correctly, only `UVC_QUIRK_FIX_BANDWIDTH` should be need to be applied, therefore only `128`.
The FPS value should not increase that much, as I already noted above. For the resolution it should work ig. Some of the crew will do testing internally soon, then I will have own results on this solution. I'm currently solely relying on the messages of looxonline.

As a note about camera-streamer, you can try to change the format there too: `custom_flags: --camera-format=YUYV`
I somehow forgot that the CM4 got camera-streamer support. I added it to the above message too.

**Reply by dragz** [reply]
Still no improvement on my setup. I can confirm that the custom_flags setting works for camera-streamer, but I cannot run both cameras at a resolution higher than 640x480. Anyway, I came to realize that as my second cam only watches the spool I don't need full resolution on it. I adjusted it down to 320x240 and fps=1 and the primary cam up to 1280x1024 and fps=20.  It works both with and without the quirks and custom_flags settings.

Again, thanks for all the effort you put into tracking this down.

**Comment by mryel00** [comment]
I got some test results from a crew member. He tried the proposed fix, but it seems like after applying it, he isn't able to use `YUYV` encoding anymore and is getting `CAP: Can't start capturing: Input/output error` or `Mainloop select() error: Success` when using `custom_flags: --format=YUYV`, any idea @looxonline? Did you come across something similar maybe?

Furthermore I discovered that ustreamer isn't working on the screenshots of @looxonline. Did you look at the streams when testing the quirk, or only at the terminal data?

**Reply by dragz** [reply]
I had to use the "Adaptive MJPEG streamer" when testing ustreamer.

**Reply by mryel00** [reply]
Fyi, these 2 errors appear without opening up any stream and force ustreamer to restart every second.

**Comment by gilbertorconde** [comment]
Im not too sure about how delayed am i to this conversation xD but i have to say, this really worked for me. I've the M8P with CM4 and im testing it with two identical Ov5640 i got from aliexpress (Z-Star Microelectronics Corp. HBVCAM FHD CAMERA). Before this i tried absolutley every possible combination between resolution, format and fps. Even tryng like 320x240 at 10fps on MJPG only one camera would work at the time. with this quirk now i have both (rather low, but still working). Now i have one of them at 640x480@30 in MJPG format and the other at 320x240@30 in YUYV

**Comment by bigtreetech** [comment]
Glad to hear that the research had a positive benefit! It really is a daft thing about the USB implementation of cameras. Would be far better for them to declare the used bandwidth rather than the bandwidth that could possible used in the absolute worst case scenario over that channel.
